Add incremental build support to gradle plugin

anatoliy-balakirev opened the pull request:
In order to make it working - I had to define clear input and output
params for each gradle task.
As part of this task I've also cleaned up some services to define clear
inputs instead of generic `ContractVerifierConfigProperties`.inputs instead of generic `ContractVerifierConfigProperties`.

Fixes gh-1133

pivotal-issuemaster commented:
@anatoliy-balakirev Please sign the [Contributor License Agreement](https://cla.pivotal.io/sign/spring?repositoryId=spring-cloud/spring-cloud-contract&pullRequestId=1165)!

[Click here](https://cla.pivotal.io/sync/spring?repositoryId=spring-cloud/spring-cloud-contract&pullRequestId=1165) to manually synchronize the status of this Pull Request.

See the [FAQ](https://cla.pivotal.io/about) for frequently asked questions.

pivotal-issuemaster commented:
@anatoliy-balakirev Thank you for signing the [Contributor License Agreement](https://cla.pivotal.io/sign/spring?repositoryId=spring-cloud/spring-cloud-contract&pullRequestId=1165)!

shanman190 commented:
So I'll admit that I've had similar issues while working through my refactor of this problem as well. Like you've seen the global mutable state either via the extension updates from within tasks or via the `ext` of the tasks.

This plugin really needs to have a pretty major refactoring to bring it up to what would be considered current standards. You've definitely gone a pretty substantial way there. 

Some things that I'd like to note as pieces we should also consider:

- Plugins are recommended not to automatically create publications unless they are very opinionated about the lifecycle. (Eg: java-gradle-plugin)
- `afterEvaluate` in this plugins case shouldn't really be necessary. The one place I've found that isn't the case is when attaching as a depends on relationship with the `compileTestJava` or `compileTestGroovy` tasks.
- Extension properties can take advantage of being lazy configuration which means that when the property is set by the user later on, then it Cascades through to the tasks that use it as well.
- Could potentially thing forward a little further to build caching. This part I'm working through ideas in relationship to how the `StubsDowloader` downloads the contracts to a temporary directory outside of Gradle's lifecycle and is additionally managing this temp storage with the JVM shutdown cleanup in `TempStorage`. If we let this just be temp storage and copy to an output then it shouldn't be so bad.

These are just some of the things that are coming to mind currently. I'll try to get my branch that I'm working on cleaned up and pushed out for a compare and contrast if anybody is interested?

anatoliy-balakirev commented:
@shanman190 oh, so you worked on the same in parallel? You sould've mentioend this in the issue, could save us quite some time...

Regarding your points:
1. Interesting, wasn't aware of this (I'm quite new to gradle plugin development; this is actually my first one :) ). But I'm not sure this is related to the incremental build feature, I worked on originally. Maybe worth to create a separate issue / PR for that?
2. and 3. - Nice! Wasn't aware that lazy evaluation can be used with extension properties. Will try to adjust that. I used `afterEvaluate` to actually simulate this laziness in that particular part.
4. Good idea! But again, maybe worth a separate issue / PR?

marcingrzejszczak commented on spring-cloud-contract-stub-runner/src/main/java/org/springframework/cloud/contract/stubrunner/ContractDownloader.java:
```diff
@@ -62,22 +61,24 @@ public ContractDownloader(StubDownloader stubDownloader,
+	 * the contracts for all the projects. We're interested only in its subset.
 	 * @return location of the unpacked downloaded stubs
 	 */
-	public File unpackedDownloadedContracts(ContractVerifierConfigProperties config) {
```
It's a public method, we should deprecate it instead of removing it.

anatoliy-balakirev replied:
@marcingrzejszczak where is this rule applicable? I mean many classes are used only internally even if methods are public, so if I modify all usages within the project - do I still need to preserve an old method? Is there some module, which is considered public API, which must follow this rule?

marcingrzejszczak replied:
In general if a class is public people might reference it. That's why wherever I can I'm using the package scoped classes. Since this is a library, I have no knowledge of who is actually using this class and how.

> Is there some module, which is considered public API, which must follow this rule?

I'm assuming that all of the public classes follow that rule. At least I try to assume so ;)

anatoliy-balakirev replied:
Oh, that's painful... I mean in general, people can use some class / method even if it's not public :) (reflection). So I thought it's usually good to draw line somewhere and say that these classes are part of the public API and are designed to be used externally and those are internal things / implementation details and can be modified. Otherwise refactoring is just impossible.

marcingrzejszczak replied:
> Oh, that's painful... I mean in general, people can use some class / method even if it's not public :) (reflection)

If someone uses reflection then there's not much I can do ;)

>  So I thought it's usually good to draw line somewhere and say that these classes are part of the public API and are designed to be used externally and those are internal things / implementation details and can be modified. Otherwise refactoring is just impossible.

That's a valid point. That's why we try not to have public API as much as possible.

marcingrzejszczak commented on spring-cloud-contract-stub-runner/src/main/java/org/springframework/cloud/contract/stubrunner/ContractDownloader.java:
```diff
@@ -62,22 +61,24 @@ public ContractDownloader(StubDownloader stubDownloader,
+		return unpackedContractStubs.getValue();
 	}
 
-	public ContractVerifierConfigProperties updatePropertiesWithInclusion(
```
It's a public method, we should deprecate it instead of removing it.

marcingrzejszczak commented on spring-cloud-contract-stub-runner/src/test/groovy/org/springframework/cloud/contract/stubrunner/ContractDownloaderSpec.groovy:
```diff
@@ -32,36 +30,31 @@ class ContractDownloaderSpec extends Specification {
 		given:
-			String contractPath = File.separator + ['a', 'b', 'c', 'd'].join(File.separator)
-			ContractDownloader contractDownloader = new ContractDownloader(stubDownloader,
+			def contractPath = File.separator + ['a', 'b', 'c', 'd'].join(File.separator)
```
I definitely prefer to have those typed :)

anatoliy-balakirev replied:
@marcingrzejszczak I saw multiple styles of using this within the project: sometimes it's `def`, sometimes exact type. `def` is quite concise, so I used it here. What is the rule in general? I'm not so familiar with groovy, but I thought that type will actually be derived based on the assigned value (i.e. similar to `var` in the latest java, except for that you can reassign it). That seems to be not so bad... Do you prefer to define concrete types everywhere instead?

marcingrzejszczak replied:
Yeah, so I prefer those to be explicit always. I know that they are not done in such a way everywhere, but I try to make those explicit. Groovy is a dynamic language so sometimes the compiler gets lost on what the type is.

marcingrzejszczak commented on spring-cloud-contract-tools/spring-cloud-contract-gradle-plugin/src/main/groovy/org/springframework/cloud/contract/verifier/plugin/ContractsCopyTask.groovy:
```diff
@@ -16,18 +16,21 @@
+import org.gradle.api.file.Directory
+import org.gradle.api.file.DirectoryProperty
+import org.gradle.api.provider.Provider
+import org.gradle.api.tasks.*
```
Let's be explicit about the imports please

anatoliy-balakirev replied:
Was automatically squashed by IDE. Will fix.

marcingrzejszczak commented:
> So I'll admit that I've had similar issues while working through my refactor of this problem as well. Like you've seen the global mutable state either via the extension updates from within tasks or via the ext of the tasks.

Yup, that's a problem.

> Plugins are recommended not to automatically create publications unless they are very opinionated about the lifecycle. (Eg: java-gradle-plugin)

I guess we're opinionated. We want to have the stub jar publication

marcingrzejszczak commented:
> In some other places it's used as an output. I.e. nothing is taken from it, but something is modified in it for further usages (which is not always the case as we are not always sharing the same instance, which is making it even harder to reason about). The fact that this class is mutable, makes this scenario possible everywhere in the code and makes analysing the flow quite hard.

I agree so we can convert the inputs and outputs separately from removing that property configuration.

> Modify as much services as possible to be very explicit about what they need (and there are also some TODOs in the code, suggesting this, which actually encouraged me to proceed);

I dislike being extremely explicit cause afterwards, if there's one more field needed to be used I need to refactor a milion places. Also, often, I have to deprecate an existing public constructor and create a new one. Remember that if we have a public API then we can't just remove stuff. We need to deprecate it and remove it when we have version e.g. 3.0.0

Even if we decide to go that way, I'd prefer that to be done in a separate PR.

> Leave existing services as is (i.e. using that global variable) and only change gradle plugin to dynamically create a new instance of that class every time I need to interact with those services. I was considering this option as well as a bit easier to start with.

Let's go with that if possible. If you think that option 2 makes more sense and won't interfere with the public API then let's go with option 2.

> It is not clear whether service is mutating something inside the ContractVerifierConfigProperties and that should be used as an output;

We can try to find those points and first change the process somehow and only then change the API?

WDYT?

anatoliy-balakirev commented:
> I dislike being extremely explicit cause afterwards, if there's one more field needed to be used I need to refactor a milion places. 
Yes, there is always some tradeoff... It's either easier to add something without changing many places, but then harder to read / understand what we ended up having, or harder to modify, but easier to reason about each service's API. As at the end developers are spending much more time reading the code than actually writing it - I would still go for the approach with properly defined APIs, TBH.

Anyway, let me try to roll back my `TestGeneratorInput` and changes to public APIs to have smaller PR for now.

anatoliy-balakirev commented on spring-cloud-contract-stub-runner/src/main/java/org/springframework/cloud/contract/stubrunner/ContractDownloader.java:
```diff
@@ -162,19 +189,6 @@ private String wrapWithAntPattern(String path) {
 				+ "**/";
 	}
 
-	private File unpackAndDownloadContracts() {
```
Just moved this method up as made it public.

anatoliy-balakirev commented on spring-cloud-contract-stub-runner/src/main/java/org/springframework/cloud/contract/stubrunner/ContractDownloader.java:
```diff
@@ -70,14 +70,43 @@ public ContractDownloader(StubDownloader stubDownloader,
 	 * @return location of the unpacked downloaded stubs
 	 */
+	// Use unpackAndDownloadContracts() and createNewInclusionProperties() instead
+	@Deprecated
```
These two methods are modifying `ContractVerifierConfigProperties`, so I deprecated them in favour of more straightforward ones (tow separate calls).

anatoliy-balakirev commented on spring-cloud-contract-tools/spring-cloud-contract-converters/src/main/groovy/org/springframework/cloud/contract/verifier/converter/RecursiveFilesConverter.groovy:
```diff
@@ -43,27 +43,40 @@ import org.springframework.cloud.contract.verifier.wiremock.DslToWireMockClientC
+	private final boolean excludeBuildFolders
 
+	// Use constructor without ContractVerifierConfigProperties
+	@Deprecated
```
Decided to keep my changes here to only pass what is actually needed, so deprecating these old constructors.

anatoliy-balakirev commented on spring-cloud-contract-tools/spring-cloud-contract-gradle-plugin/src/main/groovy/org/springframework/cloud/contract/verifier/plugin/GenerateServerTestsTask.groovy:
```diff
@@ -16,99 +16,113 @@
+		logger.info("Generated test sources dir [${generatedTestSourcesDir}]")
+		logger.info("Generated test resources dir [${generatedTestResourcesDir}]")
+		// TODO: Double check this. `ContractDependency` seems to always be `true`. I believe it was wrong here:
+		// if (getConfigProperties().getContractDependency()) {
```
Need a feedback on this. Looks like it was a bug here, but I'm not 100% sure...

marcingrzejszczak replied:
Yeah it looks like it. We should check if either group id of a dependency is set or its string notation.

anatoliy-balakirev replied:
@marcingrzejszczak question is: do we really need to do this here? If you check, `GradleContractsDownloader` is updating that `includedContracts` property to some calculated value in case there is group id / we downloaded it. Then this property is only used here. And if in exactly the same scenario we reset this property back to default value, then whole idea of calculating it in the `GradleContractsDownloader` is not clear to me. Moreover, maven plugin doesn't seem to have logic like this.

marcingrzejszczak replied:
Absolutely we don't need to do it here. Actually, we already do it in the `GradleContractsDownloader` as you say, so this shouldn't even be here.

anatoliy-balakirev replied:
Ok, removed.

anatoliy-balakirev commented:
@marcingrzejszczak I've reverted all my changes related to `TestGeneratorInput`. Still, I left my other changes to a couple of services, switching them to more fine grained input. When services are public - left an old signature as well, deprecating it. Can you take a look pls? Is it better now or you would rather do those changes in the separate PR as well (they seem to be quite not big, so decided to leave).

anatoliy-balakirev commented on spring-cloud-contract-stub-runner/src/main/java/org/springframework/cloud/contract/stubrunner/ContractDownloader.java:
```diff
@@ -189,4 +203,26 @@ private String slashSeparatedGroupId() {
 		return this.projectGroupId.replace(".", File.separator);
 	}
 
+	public static class InclusionProperties {
```
Will add.

anatoliy-balakirev replied:
Added

anatoliy-balakirev commented on spring-cloud-contract-stub-runner/src/main/java/org/springframework/cloud/contract/stubrunner/ContractDownloader.java:
```diff
@@ -70,14 +70,43 @@ public ContractDownloader(StubDownloader stubDownloader,
 
+	// Use createNewInclusionProperties() instead
+	@Deprecated
 	public ContractVerifierConfigProperties updatePropertiesWithInclusion(
```
Sorry, I'm not sure I understand what you mean. Can you elaborate pls?

marcingrzejszczak replied:
You've deprecated the old method so maybe we can start calling the new one from the old one

anatoliy-balakirev replied:
Ah, ok. But I'm doing exactly that...

marcingrzejszczak replied:
Ok I'm blind then ;)

anatoliy-balakirev commented on spring-cloud-contract-stub-runner/src/main/java/org/springframework/cloud/contract/stubrunner/ContractDownloader.java:
```diff
@@ -70,14 +70,43 @@ public ContractDownloader(StubDownloader stubDownloader,
+	@Deprecated
 	public ContractVerifierConfigProperties updatePropertiesWithInclusion(
 			File contractsDirectory, ContractVerifierConfigProperties config) {
+		final InclusionProperties newInclusionProperties = createNewInclusionProperties(contractsDirectory);
```
Let's discuss above. I believe this is exactly what I'm doing, but maybe I'm missing something :)

shanman190 commented:
> > Plugins are recommended not to automatically create publications unless they are very opinionated about the lifecycle. (Eg: java-gradle-plugin)

> I guess we're opinionated. We want to have the stub jar publication

So on this one, I've got a story for you. In the case that I have, I'd really honestly go down the path of deprecating this and forcing the end user to specifically add the stubsJar as an artifact that they publish themselves.

Story time...

In our environment (my workplace), we actually have to have every project disable the publications as a result of an oddity we encountered and have just had time to workaround, but not really discuss until now. Basically, our build server is only allowed to push artifacts into our Artifact repository, but not allowed to delete. In several cases, we have teams that are using SCC in the form of CDC as part of the service application Gradle module. This module publishes a Spring Boot jar, source jar, etc. In our case, the stubsJar publication actually conflicts with our normal `maven-publish` publication which had our repository not been configured for write once, then it would have overwritten the root publication. In our case, we just disabled the built-in publishing and added the stubsJar as an artifact to our publication which achieves the same goal without messing up the original pom file.

```
publishing {
    publications {
        maven(MavenPublication) {
            from components.java

            // other artifacts, etc
            artifact sourceJar

            artifact stubsJar
        }
    }
}
```

Now this definitely doesn't need to be a part of this pull request in order to achieve task caching, but I think that it's something to talk about. @marcingrzejszczak, I can write up an issue if you'd prefer to handle it that way or to keep it as a reminder?

anatoliy-balakirev commented:
Something weird is going on with the build after I synced with original repo...
On the CI (https://circleci.com/gh/spring-cloud/spring-cloud-contract/8633?utm_campaign=vcs-integration-link&utm_medium=referral&utm_source=github-build-link)
Build is constantly failing with a lot of errors:
```
[ERROR] Errors: 
[ERROR]   JavaContractConverterSpec.should convert Java DSL with REST to DSL for [#contractFile]:68 NoSuchElement
[ERROR]   JavaContractConverterSpec.should convert Java DSL with REST to DSL for [#contractFile]:68 NoSuchElement
[INFO] 
[ERROR] Tests run: 1093, Failures: 5, Errors: 2, Skipped: 0
[INFO] 
```
While locally it all looks good:
```

@marcingrzejszczak I see those are new tests added by you today. Any idea what's going on there?
[INFO] 
[INFO] Results:
[INFO] 
[INFO] Tests run: 1093, Failures: 0, Errors: 0, Skipped: 0
[INFO] 
```
Moreover, reported failures are in the tests I didn't even touch.

shanman190 commented on spring-cloud-contract-tools/spring-cloud-contract-gradle-plugin/src/main/groovy/org/springframework/cloud/contract/verifier/plugin/ContractsCopyTask.groovy:
```diff
@@ -16,115 +16,111 @@
-	private File outputFolder(String root, String suffix) {
-		return getExtension().stubsOutputDir != null ?
-				project.file("${getExtension().stubsOutputDir}/${root}/${suffix}") :
+	static File createTaskOutput(String root, File stubsOutputDir, Project project, String suffix) {
```
This appears unused.

anatoliy-balakirev replied:
It's used in the `SpringCloudContractVerifierGradlePlugin`

shanman190 commented on spring-cloud-contract-tools/spring-cloud-contract-gradle-plugin/src/main/groovy/org/springframework/cloud/contract/verifier/plugin/GenerateClientStubsFromDslTask.groovy:
```diff
@@ -16,66 +16,51 @@
-			project.logger.error("Couldn't retrieve the contract dsl property set by the copy contracts task", e)
-			return props.contractsDslDir
-		}
+	static File createTaskOutput(String root, File stubsOutputDir, Project project) {
```
This appears unused.

anatoliy-balakirev replied:
The same here. Used in the `SpringCloudContractVerifierGradlePlugin`

shanman190 commented on spring-cloud-contract-tools/spring-cloud-contract-gradle-plugin/src/main/groovy/org/springframework/cloud/contract/verifier/plugin/GenerateServerTestsTask.groovy:
```diff
@@ -16,99 +16,108 @@
+	static final String TASK_NAME = 'generateContractTests'
 
-	File generatedTestSourcesDir
+	@InputDirectory
```
If you wanted to make these easier to consume/configure, you can use the `@Nested` qualifer and an object to hold them.

anatoliy-balakirev replied:
Sorry, can you elaborate a bit, pls? How do you think it'll make it easier to configure if I introduce some additional object?

shanman190 replied:
So kind of along the same lines as what @marcingrzejszczak was talking about by making a class know about a large number of inputs. The same holds true for tasks. While it's most definitely doable, it blurs the api surface. You can kind of see an example of a way to deal with it here:

https://github.com/shanman190/spring-cloud-contract/commit/34bf9af5eac2822442cb21c90c4960e9d3188d33#diff-b998e42b8cb2cccb4ff11ccc92d5f714

anatoliy-balakirev replied:
Ok, thanks. Will try to clean it up a bit.

anatoliy-balakirev replied:
Cleaned up.
Also tried to switch `ContractVerifierExtension` to be property based, but got some weird issues there and rolled back. Might be a task for some next PR. Not sure will have more time for this now and it's just a nice to have at the end, not a strict requirement.

anatoliy-balakirev replied:
Actually, it looks like I'll manage to use properties in the `ContractVerifierExtension` at the end. Already have something working. Only need to fix tests.

shanman190 commented on spring-cloud-contract-tools/spring-cloud-contract-gradle-plugin/src/main/groovy/org/springframework/cloud/contract/verifier/plugin/GradleContractsDownloader.groovy:
```diff
@@ -26,73 +25,69 @@ class GradleContractsDownloader {
-				.withProperties(extension.contractsProperties)
-		if (extension.contractRepository.proxyPort) {
-			options = options.withProxy(extension.contractRepository.proxyHost, extension.contractRepository.proxyPort)
+	static File contractsSubDirIfPresent(File contractsDirectory, Logger logger) {
```
I would think this wouldn't need to be static and probably should be private.

anatoliy-balakirev replied:
Used in the `SpringCloudContractVerifierGradlePlugin`

shanman190 replied:
Ahh, I see now. So something to note about this, is that usually you want the outputs to be a directory within the build folder since:

1) They're gitignored, so changing the data won't result in cascading files into a commit where that doesn't make sense.
2) Further cleans up the input/output relationship to the task. 
2a) You shouldn't need to let the task know about the directory conventions once you're using the lazy configuration API. Essentially, the extension could define the intermediate output directory that contracts should be placed in before being processed by the generation tasks. That output directory would then be the input to the generation tasks.

anatoliy-balakirev replied:
So, the logic here is:
1. We have an `init` task (used to be part of the `copy contracts`, but I extracted it to the dedicated one to make it incremental). That task is:
1.1 If configured, downloading and unpacking contracts into some temporary directory, which is then used as an input by `copy contracts` task. In this case, that output directory is calculated dynamically by the task itself and then we pick subdirectory from there using this `contractsSubDirIfPresent` method;
1.2 If downloading is not needed - just using default `contracts` directory within `resources` as an input for `copy contracts` task. Again, `contractsSubDirIfPresent` is used there to pick subdirectory if present. That default directory is kind of static and initialised in the `SpringCloudContractVerifierGradlePlugin` as a default output;
2. Copy contracts task is taking that directory from `init` task and copying those contracts to the `build` folder.

So it kind of does what you suggested, but not in the `init` task. Most of the times that `init` task does nothing, actually.

anatoliy-balakirev replied:
I actually moved those properties to the internal nested config class and clean this part up a bit.

shanman190 commented on spring-cloud-contract-tools/spring-cloud-contract-gradle-plugin/src/main/groovy/org/springframework/cloud/contract/verifier/plugin/StubRunnerOptionsFactory.groovy:
```diff
@@ -0,0 +1,28 @@
+import org.springframework.cloud.contract.stubrunner.spring.StubRunnerProperties
+
+@CompileStatic
+class StubRunnerOptionsFactory {
```
Might want to make this `@PackageScoped` that way it doesn't become a part of the public API, huh @marcingrzejszczak?

anatoliy-balakirev replied:
Done, thx.

shanman190 commented:
@anatoliy-balakirev, here's my work in progress if you want to use some of it for inspiration. Mine is going way left field in terms of changes, so yours will make more sense in the short term.

https://github.com/shanman190/spring-cloud-contract/commit/34bf9af5eac2822442cb21c90c4960e9d3188d33

anatoliy-balakirev commented:
@shanman190 Nice, will take a look. Thanks!

marcingrzejszczak commented:
>> > Plugins are recommended not to automatically create publications unless they are very opinionated about the lifecycle. (Eg: java-gradle-plugin) ...

@shanman190 hey thanks for the story :)

Let's create a separate issue. Removing that default would be a breaking change :/

marcingrzejszczak commented:
>Something weird is going on with the build after I synced with original repo... ...

@anatoliy-balakirev eee no idea. Our builds work fine in jenkins. Are your sure you synced with the latest changes?

anatoliy-balakirev commented:
@marcingrzejszczak I created another small PR to fix a couple of NPEs and it fails with the same:
https://circleci.com/gh/spring-cloud/spring-cloud-contract/8646?utm_campaign=vcs-integration-link&utm_medium=referral&utm_source=github-build-link

PR is here: https://github.com/spring-cloud/spring-cloud-contract/pull/1170

Both that one and this are based on the latest master, so there is definitely some issue with it / its config for CI:
![Screen Shot 2019-08-13 at 10 44 30](https://user-images.githubusercontent.com/10894915/62927772-8506f000-bdb7-11e9-9cea-9bcc3d5427af.png)

Also it started failing when I rebased it yesterday.

marcingrzejszczak commented:
Can you rebase against master and try to build it again with the new base image for circleci?

marcingrzejszczak commented:
Jacoco was the culprit. Please rebase and it should work fine.

anatoliy-balakirev commented:
@marcingrzejszczak that helped, thanks!

anatoliy-balakirev commented:
@marcingrzejszczak @shanman190 I've updated PR with all the latest changes. Switched extension to use lazy properties as well. Take a look pls, I believe it should be the final version.

anatoliy-balakirev commented:
The last rebase was to restart CI tests as they seem to be flaky.

marcingrzejszczak commented:
Congrats @anatoliy-balakirev and @shanman190 on this fantastic PR. Great work!

anatoliy-balakirev commented:
Thanks! :)

anatoliy-balakirev commented:
@marcingrzejszczak I just realised that this PR introduced some kind of backward incompatible change in terms of configuring plugin in the `build.gradle.kts` (from here: https://docs.gradle.org/current/userguide/lazy_configuration.html):
```
Note that Gradle Groovy DSL will generate setter methods for each Property-typed property in a task implementation. These setter methods allow you to configure the property using the assignment (=) operator as a convenience.

Kotlin DSL conveniences will be added in a future release.
```

So when using Kotlin DSL, instead of writing this:
```
contracts {
    packageWithBaseClasses = "com.numbrs.referral.controller.contract"
    basePackageForTests = packageWithBaseClasses
    testMode = TestMode.WEBTESTCLIENT
    testFramework = TestFramework.JUNIT5
}
```

We'll have to do this (note `.set()` instead of `=`):
```
contracts {
    packageWithBaseClasses.set( "com.numbrs.referral.controller.contract")
    basePackageForTests.set(packageWithBaseClasses)
    testMode.set(TestMode.WEBTESTCLIENT)
    testFramework.set(TestFramework.JUNIT5)
}
```

Groovy DSL is not affected.
I can create a new PR to add explicit getters if that's an issue.

marcingrzejszczak commented:
Yeah that's exactly what we had to do here https://github.com/spring-cloud/spring-cloud-contract/commit/3c4772bb6f0dcba6bdfd484657943cb539557e02

It's not a breaking change though since there was no support whatsover for Kotlin. However having consistent behavior would be great. Sure, file another PR :)

anatoliy-balakirev commented:
I checked this further and looks like currently it's not really possible to do, as Kotlin doesn't support properties with overloaded setters. Here is an original discussion:
https://github.com/gradle/kotlin-dsl/issues/597

And a couple more related links:
https://github.com/gradle/kotlin-dsl/issues/380
https://youtrack.jetbrains.com/issue/KT-24831
https://youtrack.jetbrains.com/issue/KT-4075

So for now Kotlin users will have to write `property.set("value")` instead of `property = "value"`.
So you can adjust it here as well: https://github.com/spring-cloud/spring-cloud-contract/commit/3c4772bb6f0dcba6bdfd484657943cb539557e02

That big assignment (`packageWithBaseClasses = objects.property(String::class).value("com.example.fraud")`) is not needed and can / should be replaced with:
`packageWithBaseClasses.set("com.example.fraud")`
